zstd: share context buffers between compressor instances

michoecho opened the pull request:
Every stream from/to a compressed SSTable has its own instance of compressor. In zstd's case, each compressor has its own buffer for the compression and decompression context. These buffers can be quite large -- their typical sizes are between 100 KiB and 1 MiB range, so they provide a stall hazard and an OOM hazard, given that thousands of concurrent SSTable reads are possible.

This patch fixes that hazard by making all ZSTD compressors use using a single buffer for all ZSTD compressors. This is somewhat risky, because the buffer is resized up when needed, but is never resized down. If a huge buffer appears, e.g. due to malformed compression settings, it will stay in memory forever (until node restart), even if compression settings get fixed. To prevent this, readers which need abnormally large buffers do not use the shared buffer, but use a private one.

Fixes #11733.

nyh commented:
@michoecho I don't understand something: The fact that you now reuse the same buffer for further compression tasks, means that you must (although I didn't see where it is...) already have serialization for these compression tasks - otherwise multiple compressions can end up overwriting the same buffer at the same time and everything will break.
But if you already have serialization of these compression tasks, why did we have this problem in the first place - because you said that the problem was that many of them were running in parallel?
What am I missing?

michoecho commented:
> @michoecho I don't understand something: The fact that you now reuse the same buffer for further compression tasks, means that you must (although I didn't see where it is...) already have serialization for these compression tasks - otherwise multiple compressions can end up overwriting the same buffer at the same time and everything will break. But if you already have serialization of these compression tasks, why did we have this problem in the first place - because you said that the problem was that many of them were running in parallel? What am I missing?

There are no "compression tasks". Compression is a simple synchronous function call. But it needs a memory buffer, which for performance reasons is reused between calls, rather than allocated on every call.

But currently, every SSTable reader/writer has its own buffer (when the SSTable is zstd-compressed), which lives as long as the reader. There can be thousands of readers in parallel, so this can blow up and cause OOMs. But there's no reason to have a separate buffer per each reader/writer — you can have just one for the entire program. And this is what the patch does.

nyh commented on zstd.cc:
```diff
@@ -75,34 +99,47 @@ zstd_processor::zstd_processor(const opt_getter& opts)
+    if (!_shared_dctx) {
+        auto dctx_size = ZSTD_estimateDCtxSize();
+        _shared_dctx = allocate_aligned_buffer<char>(dctx_size, 8);
+        if (!ZSTD_initStaticCCtx(_shared_dctx.get(), dctx_size)) {
```
I think there's a typo here, unless I'm minunderstanding something. Shouldn't this be calling **DC** not *CC* here?

michoecho replied:
Yes, it's a typo. Apparently the init functions mostly just zero some stuff, so it somehow worked. You just saved me from a possibly very sneaky bug, thanks.

michoecho replied:
By the way, there was also a similar typo before the patch in this place:
```
    _dctx = ZSTD_initStaticDCtx(_dctx_raw.get(), dctx_size);
    if (!_cctx) {
```
instead of:
```
    _dctx = ZSTD_initStaticDCtx(_dctx_raw.get(), dctx_size);
    if (!_dctx) {
```
though this `if` should never be taken either way.

I did my best not to repeat such a typo, but well, here we are...

nyh commented on zstd.cc:
```diff
@@ -75,34 +99,47 @@ zstd_processor::zstd_processor(const opt_getter& opts)
 
 size_t zstd_processor::uncompress(const char* input, size_t input_len, char* output, size_t output_len) const {
-    auto ret = ZSTD_decompressDCtx(_dctx, output, output_len, input, input_len);
+    auto dctx = reinterpret_cast<ZSTD_DCtx*>(_shared_dctx.get());
```
I'm curious why you dropped the "_dctx_raw" vs "_dctx" distinction in the new version of the code, forcing you to do reinterpret_casts below. What's the benefit?

michoecho replied:
No particular reason. It's just that the buffers can be reallocated now, so the typed "observer" pointers would now have to be refreshed / checked for invalidation before use. I thought that reducing the number of variables and always using the shared pointers would be simpler.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2811/

michoecho commented:
v2:
- Fixed the typo bug noticed by @nyh.
- Made allocation of compression contexts lazy. (Moved it from `compressor` constructor to compress()). Rationale explained in comments in get_cctx().
- Removed the new test case in sstables_3_x_test. It was unreasonably slow in debug mode.
- Added a separate test file with faster and more focused tests.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2817/

avikivity commented on sstables/compress.cc:
```diff
@@ -33,7 +33,7 @@ enum class mask_type : uint8_t {
 // size_bits cannot be >= 64
 static inline uint64_t make_mask(uint8_t size_bits, uint8_t offset, mask_type t) noexcept {
-    const uint64_t mask = ((1 << size_bits) - 1) << offset;
+    const uint64_t mask = ((uint64_t(1) << size_bits) - 1) << offset;
```
Note if it does happen, the server is dead, since it means we allocate 4GB. A shard usually only has 7GB.

michoecho replied:
I don't know what this code is responsible for and I wasn't able to understand it with a cursory look. But somehow this triggered UBSAN in the CI for v1 of this patch (https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2811/), and the node apparently didn't die.

michoecho replied:
Clarification: UBSAN was triggered by a new test, which used huge compression chunks. Not by the `zstd.cc` patch itself.

avikivity replied:
A test wouldn't kill the node, but a production node that suddenly tried to allocate 4GB would die a horrible death.

michoecho replied:
I think you are confusing this piece of code with something else. AFAIK the numbers here aren't related to any allocation sizes.

As far as I understand, this code is used for bitpacking offsets of compression chunks in the data file. In particular, in the case which triggers UBSAN `size_bits` is a bit width of a file offset. So >32 is perfectly possible here. Nothing allocation-related.

michoecho replied:
I don't understand exactly what's being encoded here, but I know that the voodoo magic in https://github.com/scylladb/scylladb/blob/b9d88a36011b88788df82e2d2d5036a8e6fec298/sstables/segmented_compress_params.hh says that all you need to trigger this UB is to set the compression chunk size to >=2MiB.

michoecho replied:
The 2MiB bound follows from this line: https://github.com/scylladb/scylladb/blob/b9d88a36011b88788df82e2d2d5036a8e6fec298/sstables/segmented_compress_params.hh#L49

avikivity replied:
Ah, you're right, that's part of CompressionInfo.db management.

avikivity replied:
@denesb please verify, you wrote the code. There's no question the fix is correct, just if we need to backport it. Likely not, since we'd have seen a bug here long ago.

denesb replied:
I wrote this a lifetime ago. The size passed to this method is always `_segmented_offset_size_bits`, calculated here:
```
https://github.com/scylladb/scylladb/blob/b9d88a36011b88788df82e2d2d5036a8e6fec298/sstables/compress.cc#L171
```
After plugging in various chunk sizes I found that we can indeed get a value of 33 with the largest supported chunk size of 2**30 (1GiB). But the one below that (512MiB) is already below this.

michoecho replied:
> the size passed to this method is always _segmented_offset_size_bits

@denesb That's not true. Sometimes the size passed to this method is `_segment_base_offset_size_bits`, which is calculated using the table at https://github.com/scylladb/scylladb/blob/b9d88a36011b88788df82e2d2d5036a8e6fec298/sstables/segmented_compress_params.hh#L31.

> that we can indeed get a value of 33 with the largest supported chunk size of 2**30 (1GiB). But the one below that (512MiB) is already below this.

The table I linked dictates a relationship between chunk size and `_segment_base_offset_size_bits`. You can see that for a chunk size with bit width >=21, (meaning chunk size >=  2MiB), `_segment_base_offset_size_bits` is >=32, triggering the UB. So the problem starts with 2MiB, not 512MiB.

You can easily verify that's the case by setting the compression chunk size to 2MiB in some test and running it in debug mode. It will trip UBSAN.

michoecho replied:
(I did run such test.)

denesb replied:
Right, I missed the usage in `write()`. Well, then there is not much to discuss, we have to backport this. I wonder how no one hit this yet.

michoecho replied:
>  I wonder how no one hit this yet.

2MiB chunks are pretty big. Maybe nobody uses them?

I also don't see any tests which use 2 MiB chunks. I think the biggest chunk size we test is 1 MiB, in SCT: https://github.com/scylladb/scylla-cluster-tests/blob/bfb8ebf684134257de920f4e7d97eb834716d91a/sdcm/nemesis.py#L2260

michoecho replied:
But do note that with chunks <2MiB the result will be well-defined, but it still may be wrong, because information is lost in the second shift (`<< offset`).

E.g. with `size_bits == 25` and `offset == 7`, `mask` is going to be equal to `0xffffffff'ffffff80`, instead of the intended `0x0000007f'ffffff80`, due to sign extension.

denesb replied:
Wrong result from this method would cause failure to read compressed sstables. So I think it is unlikely we've hit this in production. I think it is best to write an exploratory unit test mapping up the bounds of chunk sizes, within which the function is correct and incorrect.

michoecho replied:
> So I think it is unlikely we've hit this in production.

So do I.

> I think it is best to write an exploratory unit test mapping up the bounds of chunk sizes, within which the function is correct and incorrect.

I don't understand this proposal.

denesb replied:
> I don't understand this proposal.

I'd like to have a unit test which uses compression with carious chunk sizes. In particular I want this test to answer the question of what is the smallest chunk size that already causes problems before the fix. But we should also test the smallest and largest chunk sizes to make sure we don't regress in the future either.
Maybe you already have a test which accomplishes all this, in which case there is no additional changes needed.

avikivity replied:
2MB chunks will not work well.

michoecho replied:
@denesb I think I would need to spend a considerable amount of effort to write such a test. It would require gaining understanding of the CompressionInfo code.

In particular, the test which originally uncovered this UB was still passing accidentally despite the bug, due to the exact argument values — the write and read masks were wrong, but because they are _both_ equally wrong, and the written value (0) happened to fit in these masks, the result came out OK. So the test would require ensuring that actually multiple chunks/buckets/something are written, or something like that, so that the loss of information due to bad masks is actually meaningful.

michoecho replied:
(And this may require something more involved than just running any workload with chunk sizes configured so-and-so).

michoecho replied:
The thing is, I'm not sure what values go to these masks.

E.g. if the value that goes there is always 0, there's a dreadful possibility that the bug actually happens in production and writes some garbage to compressioninfo, but the bugs in the read path and write path happen to cancel each other, and fixing the bug will cause a regression because the corrupted data will no longer be readable...

denesb replied:
You can find the design for the compression of compressioninfo here: https://github.com/scylladb/scylladb/issues/1946 (https://github.com/scylladb/scylladb/issues/1946#issuecomment-305999782).
The entire machinery is used to store in a compressed form the compression-info component of already used sstables, it is used exclusively on reads. So no chance of a mistake in `make_mask()` causing corrupt sstables. Also, you do not have to be afraid that a fix to this code will cause problems with old nodes. The in-memory state doesn't persist between restarts.

A wrong value calculated by `make_mask()` *could* go unnoticed if it resulted in looking up the wrong chunk on the disk. But this is highly unlikely, as that would almost surely would result in a parse failure, or even an earlier decompress failure.

avikivity commented on compress.cc:
```diff
@@ -22,7 +22,7 @@ class lz4_processor: public compressor {
                     size_t output_len) const override;
     size_t compress(const char* input, size_t input_len, char* output,
-                    size_t output_len) const override;
+                    size_t output_len) override;
```
I think compress() is still logically const, since it doesn't change the compressor, only the output.

This is one of the few cases where `mutable` is indicated. It's state that doesn't affect the result of computations.

avikivity commented on zstd.cc:
```diff
@@ -22,15 +22,41 @@ static const sstring COMPRESSOR_NAME = compressor::namespace_prefix + "ZstdCompr
+    // The max context size for the default compression level 3 is 1.3 MiB.
+    // The max context size for max compression level 22 and chunks of size
+    // 64 KiB is 1.8 MiB. So 2 MiB seems like a good threshold for "reasonable"
+    // setups.
```
An alternative approach is to decay the buffer sizes over time. I don't see an advantage to one over the other, 2 MiB/shard is reasonable overhead and I agree exceeding it will be rare (if traumatic).

michoecho commented:
Abandoning in favor of #13324.
